server/src/main/java/io/deephaven/server/session/SessionState.java

```diff
@@ -633,6 +634,43 @@ public boolean isNonExport() {
             return exportId == NON_EXPORT_ID;
         }
 
+        /**
+         * Applies the function {@code f} to the results of {@code this} export.
+         *
+         * <p>
+         * In the case where {@code this} export was built with a {@link SessionState}, the function {@code f} will be
+         * invoked inside {@link ExportBuilder#submit(Callable) submit} with the same session.
+         *
+         * <p>
+         * In the case where {@code this} export was not built with a {@link SessionState}, the function {@code f} will
+         * be invoked on this thread.
+         *
+         * <p>
+         * In both cases, the resulting export is a "non-export".
+         *
+         * @param f the function
+         * @return the new export after applying
+         * @param <R> the new exported type
+         */
+        public <R> ExportObject<R> map(Function<? super T, ? extends R> f) {
+            if (session == null) {
+                final T localResult = result;
```

Comment:
You first need to acquire a ref count on the export object. If you do not, then the export object may be released even if you get a non-null value for result. If the result is a liveness referent and the export object was the only liveness manager managing the result, then the result will be released even though you do not own a refcount.

See the comment on `ExportObject#get`:


```java
        /**
         * WARNING! This method call is only safe to use in the following patterns:
         * <p>
         * 1) If an export (or non-export) {@link ExportBuilder#require}'d this export then the method is valid from
         * within the Callable/Runnable passed to {@link ExportBuilder#submit}.
         * <p>
         * 2) By first obtaining a reference to the {@link ExportObject}, and then observing its state as
         * {@link ExportNotification.State#EXPORTED}. The caller must abide by the Liveness API and dropReference.
         * <p>
         * Example:
         *
         * <pre>
         * {@code
         * <T> T getFromExport(ExportObject<T> export) {
         *     if (export.tryRetainReference()) {
         *         try {
         *             if (export.getState() == ExportNotification.State.EXPORTED) {
         *                 return export.get();
         *             }
         *         } finally {
         *             export.dropReference();
         *         }
         *     }
         *     return null;
         * }
         * }
         * </pre>
         *
         * @return the result of the computed export
         */
        public T get() {
```

Technically, we should update the comment to highlight the importance of managing the result before the dropReference in the `finally` block:
```java
if (export.getState() == ExportNotification.State.EXPORTED) {
    final T localResult = export.get();
    if (localResult instanceof LivenessReferent) {
        LivenessScopeStack.peek().manage(localResult);
    }
    return localResult;
}
```

Note that this mapping implementation could skip managing the result as long as the mapping function doesn't need to own a copy. As in, table operations will manage the result where appropriate, but any function that holds a hard-reference would need to be aware that it must manage the result to ensure it remains valid for the length of time it wants result to be valid.

Whatever approach we choose should be @rcaudy approved.

Do note that session-less non-exports are valid; perhaps we should prefer that approach even if there is some extra threading involved using that code path. I would also be willing to consider a feature that allows for creation of session-less non-exports that are going to be invoked immediately. See `#scheduleExport` and how it delegates to `doExport`. This feature would add a third branch that instead calls `doExport` from within `scheduleExport`.

Though, if `this` is not already in an `EXPORTED` state, then execution must be deferred. So, maybe such a feature has limited use.